src/p4p/test/system_client_server.py

```diff
@@ -0,0 +1,631 @@
+
+hello_dict = {
+    'meta': {
+        'description': 'Hardware Block with a greet() Method',
+        'tags': [],
+        'writeable': False,
+        'label': 'TESTHELLO',
+        'fields': ['health', 'greet', 'error']
+    },
+    'health': {
+        'value': "OK",
+        'alarm': alarm_ok,
+        'timeStamp': ts_zero,
+        'meta': {
+            'description': 'Displays OK or an error message',
+            'tags': ['widget:textupdate'],
+            'writeable': False,
+            'label': 'Health'
+        }
+    },
+    'greet': {
+        'takes': {
+            'elements': {
+                'name': {
+                    'description': 'The name of the person to greet',
+                    'tags': ['widget:textinput'],
+                    'writeable': True,
+                    'label': 'Name'
+                },
+                'sleep': {
+                    'dtype': 'float64',
+                    'description': 'Time to wait before returning',
+                    'tags': ['widget:textinput'],
+                    'writeable': True,
+                    'label': 'Sleep'
+                }
+            },
+            'required': ["name"]
+        },
+        'defaults': {
+            'sleep': 0.0
+        },
+        'description': 'Optionally sleep <sleep> seconds, then return a greeting to <name>',
+        'tags': [],
+        'writeable': True,
+        'label': 'Greet',
+        'returns': {
+            'elements': {
+                'return': {
+                    'description': 'The manufactured greeting',
+                    'tags': ['widget:textupdate'],
+                    'writeable': False,
+                    'label': 'Return'
+                }
+            },
+            'required': ['return']
+        }
+    },
+    'error': {
+        'takes': empty_map_meta_dict,
+        'description': 'Raise an error',
+        'tags': [],
+        'writeable': True,
+        'label': 'Error',
+        'returns': empty_map_meta_dict
+    }
+}
+
+hello_expected = Value(hello_block_t, hello_dict)
+
+
+class Channel(object):
+    # TODO: Why is this needed?
+    Value = Value
+
+    def __init__(self, structure, field_name=None):
+        self.structure = structure
+        self.field_name = field_name
+
+    def get(self):
+        if self.field_name:
+            ret = self.structure[self.field_name]
+        else:
+            ret = self.structure
+        return ret
+
+    def monitor(self, updater):
+        # TODO: what is good here?
+        self.structure.attach(updater, self.field_name)
+
+    def rpc(self, response, request):
+        raise NotImplementedError()
+
+    def put(self, response, request):
+        raise NotImplementedError()
+
+
+class BlockChannel(Channel):
+    def __init__(self, structure, channels):
+        self.channels = channels
+        super(BlockChannel, self).__init__(structure)
+
+    def rpc(self, response, request):
+        # TODO: need the connection request object
```

Comment:
NTURI didn't exist when I wrote this so we have to support what I currently have anyway. 

Looking at it now, NTURI does most of what I want, but there are some confusing features:
* scheme: what is the point of this? The client has just connected over pva, why does it need to be told again? I would at least expect this to be optional.
* path: again, the example shows this to be the same as the channel name, although I guess it doesn't, so this could be a sub-path relative to the channel... Also, I would expect it to be optional.
* query: why limit to scalar types? I would have made this an arbitrary structure...
* <field-type>: What happens if you want an arbitrary structure called path?

I could be persuaded to use this if the extra <field-type> fields were pushed into query, then I could support exactly what I have now, but separating scalar and structure fields with the possibility of a name clash seems clumsy...

Reply:
I didn't originate NTURI, and agree that it contains some unnecessary redundancy.  I advocate it for the sole reason that the alternative is clearly everyone inventing something different (MASAR originally did as well).  This prevents the interoperability which has been at the core of why epics has been so widely adopted.

wrt. redundancy it may to know that these bits are "borrowed" from HTTP, where they exist to facility proxies.

> why limit to scalar types?

I don't see this as a real limitation.  The RPC wrappers of P4P allow arbitrary Value to be passed.

The important points which I see are that: each method has a different PV name, and the arguments are stored in a sub-structure 'query'.  And the pvRequest isn't used.

wrt. legacy.  In the case of MASAR, the new server handles calls with both scheme to maintain compatibility.  P4P has the Dispatcher concept to encapsulate this argument decoding step.

Reply:
Interesting, so I could move new code to use NTURI and the old code would still work, this is a good way forward. If I can populate query with an arbitrary (possibly nested) structure, then that would do everything I need.

Reply:
In my mind, this approach is the only realistic way forward if there are users.

If you're interested, in the MASAR storeServiceConfig() call the 'config' argument is an NTTable.  The expected definition is the same as the return type 'configTable' which is defined near the top of the file.

https://github.com/mdavidsaver/masarService/blob/a6f830d1c067046dd9355e1dc32b6b2fec9effe2/python/minimasar/ops.py#L85-L86